subsys/net/ip/net_if.c

```diff
@@ -1296,44 +1277,185 @@ static void remove_prefix_addresses(struct net_if *iface,
 	struct net_if_ipv6 *ipv6;
 
 	NET_DBG("Prefix %s/%d expired",
-		net_sprint_ipv6_addr(&prefix->prefix), prefix->len);
+		net_sprint_ipv6_addr(&ifprefix->prefix), ifprefix->len);
 
-	prefix->is_used = false;
+	ifprefix->is_used = false;
 
-	if (net_if_config_ipv6_get(prefix->iface, &ipv6) < 0) {
+	if (net_if_config_ipv6_get(ifprefix->iface, &ipv6) < 0) {
 		return;
 	}
 
 	/* Remove also all auto addresses if the they have the same prefix.
 	 */
-	remove_prefix_addresses(prefix->iface, ipv6, &prefix->prefix,
-				prefix->len);
+	remove_prefix_addresses(ifprefix->iface, ipv6, &ifprefix->prefix,
+				ifprefix->len);
+
+	net_mgmt_event_notify(NET_EVENT_IPV6_PREFIX_DEL, ifprefix->iface);
+}
+
+static void prefix_timer_remove(struct net_if_ipv6_prefix *ifprefix)
+{
+	NET_DBG("IPv6 prefix %s/%d removed",
+		net_sprint_ipv6_addr(&ifprefix->prefix),
+		ifprefix->len);
+
+	ifprefix->lifetime.timer_timeout = 0;
+	ifprefix->lifetime.wrap_counter = 0;
+
+	sys_slist_find_and_remove(&active_prefix_lifetime_timers,
+				  &ifprefix->lifetime.node);
+}
+
+static bool prefix_manage_timeout(struct net_if_ipv6_prefix *ifprefix,
+				  u32_t current_time, u32_t *next_wakeup)
+{
+	if (check_timeout(ifprefix->lifetime.timer_start,
+			  ifprefix->lifetime.timer_timeout,
+			  ifprefix->lifetime.wrap_counter,
+			  current_time)) {
+		prefix_lifetime_expired(ifprefix);
+		return true;
+	}
+
+	if (current_time == NET_TIMEOUT_MAX_VALUE) {
+		ifprefix->lifetime.wrap_counter--;
+	}
+
+	if (ifprefix->lifetime.wrap_counter > 0) {
+		*next_wakeup = NET_TIMEOUT_MAX_VALUE;
+	} else {
+		*next_wakeup = ifprefix->lifetime.timer_timeout;
+	}
+
+	return false;
+}
+
+static void prefix_lifetime_timeout(struct k_work *work)
+{
+	u64_t timeout_update = UINT64_MAX;
+	u32_t current_time = k_uptime_get_32();
+	bool found = false;
+	struct net_if_ipv6_prefix *current, *next;
+
+	ARG_UNUSED(work);
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&active_prefix_lifetime_timers,
+					  current, next, lifetime.node) {
+		u32_t next_timeout;
+		bool is_timeout;
+
+		is_timeout = prefix_manage_timeout(current, current_time,
+						   &next_timeout);
+		if (!is_timeout) {
+			if (next_timeout < timeout_update) {
+				timeout_update = next_timeout;
+				found = true;
+			}
+		}
+
+		if (current == next) {
+			break;
+		}
+	}
+
+	if (found) {
+		/* If we are near upper limit of s32_t timeout, then lower it
+		 * a bit so that kernel timeout will not overflow.
+		 */
+		if (timeout_update >= NET_TIMEOUT_MAX_VALUE) {
+			timeout_update = NET_TIMEOUT_MAX_VALUE;
+		}
+
+		NET_DBG("Waiting for %d ms", (u32_t)timeout_update);
 
-	net_mgmt_event_notify(NET_EVENT_IPV6_PREFIX_DEL, prefix->iface);
+		k_delayed_work_submit(&prefix_lifetime_timer, timeout_update);
+	}
+}
+
+static void prefix_submit_work(struct net_if_ipv6_prefix *ifprefix)
+{
+	s32_t remaining;
+
+	remaining = k_delayed_work_remaining_get(&prefix_lifetime_timer);
+	if (!remaining || (ifprefix->lifetime.wrap_counter == 0 &&
```

Comment:
This complex, duplicated logic should be factored out and reused.

Reply:
Some part of the code could be refactored and more common functions could be introduced. It just requires functions with lot of parameters etc but can be done of course.

Reply:
If it requires functions with lots of params, something is wrong. You define `struct net_timeout`, and then you would define "methods" for it, taking pointer to that structure and minimally needed params, and would mostly operate on the internal state of struct net_timeout. For example, there would be net_timeout_expired(), which, besides struct net_timeout*, would take only one param - reference to current time.